File: sdk/typescript/test/path.test.ts
import { describe, it, expect } from 'vitest';
import { parsePath, getPath, setPath, deletePath } from '../src/path.js';

describe('parsePath', () => {
  it('parses a simple key', () => {
    expect(parsePath('name')).toEqual([{ type: 'key', value: 'name' }]);
  });

  it('parses dotted keys', () => {
    expect(parsePath('a.b.c')).toEqual([
      { type: 'key', value: 'a' },
      { type: 'key', value: 'b' },
      { type: 'key', value: 'c' },
    ]);
  });

  it('parses a positive index', () => {
    expect(parsePath('items[0]')).toEqual([
      { type: 'key', value: 'items' },
      { type: 'index', value: 0 },
    ]);
  });

  it('parses a negative index', () => {
    expect(parsePath('items[-1]')).toEqual([
      { type: 'key', value: 'items' },
      { type: 'index', value: -1 },
    ]);
  });

  it('parses a wildcard', () => {
    expect(parsePath('endpoints[*].host')).toEqual([
      { type: 'key', value: 'endpoints' },
      { type: 'wildcard' },
      { type: 'key', value: 'host' },
    ]);
  });

  it('parses length as terminal', () => {
    expect(parsePath('items.length')).toEqual([
      { type: 'key', value: 'items' },
      { type: 'length' },
    ]);
  });

  it('treats length as key when it is the first segment', () => {
    expect(parsePath('length')).toEqual([{ type: 'key', value: 'length' }]);
  });

  it('parses chained brackets', () => {
    expect(parsePath('matrix[0][1]')).toEqual([
      { type: 'key', value: 'matrix' },
      { type: 'index', value: 0 },
      { type: 'index', value: 1 },
    ]);
  });

  it('parses mixed path', () => {
    expect(parsePath('folder[5].person.mobile')).toEqual([
      { type: 'key', value: 'folder' },
      { type: 'index', value: 5 },
      { type: 'key', value: 'person' },
      { type: 'key', value: 'mobile' },
    ]);
  });

  it('returns empty array for empty string', () => {
    expect(parsePath('')).toEqual([]);
  });
});

describe('getPath', () => {
  it('resolves a simple key', () => {
    expect(getPath({ name: 'alice' }, 'name')).toBe('alice');
  });

  it('resolves nested keys', () => {
    expect(getPath({ db: { host: 'localhost', port: 5432 } }, 'db.host')).toBe('localhost');
    expect(getPath({ db: { host: 'localhost', port: 5432 } }, 'db.port')).toBe(5432);
  });

  it('returns null for missing key', () => {
    expect(getPath({ a: 1 }, 'b')).toBeNull();
  });

  it('returns null for deep missing key', () => {
    expect(getPath({ a: { b: 1 } }, 'a.c.d')).toBeNull();
  });

  it('resolves positive array index', () => {
    expect(getPath({ items: ['alpha', 'beta', 'gamma'] }, 'items[0]')).toBe('alpha');
    expect(getPath({ items: ['alpha', 'beta', 'gamma'] }, 'items[2]')).toBe('gamma');
  });

  it('resolves negative array index', () => {
    expect(getPath({ items: ['alpha', 'beta', 'gamma'] }, 'items[-1]')).toBe('gamma');
    expect(getPath({ items: ['alpha', 'beta', 'gamma'] }, 'items[-2]')).toBe('beta');
  });

  it('returns null for out of bounds index', () => {
    expect(getPath({ items: ['a'] }, 'items[5]')).toBeNull();
    expect(getPath({ items: ['a'] }, 'items[-5]')).toBeNull();
  });

  it('resolves wildcard to collect values', () => {
    const data = {
      endpoints: [
        { host: 'a.com', port: 80 },
        { host: 'b.com', port: 443 },
      ],
    };
    expect(getPath(data, 'endpoints[*].host')).toEqual(['a.com', 'b.com']);
    expect(getPath(data, 'endpoints[*].port')).toEqual([80, 443]);
  });

  it('resolves length of array', () => {
    expect(getPath({ items: ['a', 'b', 'c'] }, 'items.length')).toBe(3);
  });

  it('resolves length of object', () => {
    expect(getPath({ m: { a: 1, b: 2 } }, 'm.length')).toBe(2);
  });

  it('resolves chained brackets for nested arrays', () => {
    const data = {
      matrix: [
        [1, 2, 3],
        [4, 5, 6],
      ],
    };
    expect(getPath(data, 'matrix[0][1]')).toBe(2);
    expect(getPath(data, 'matrix[1][-1]')).toBe(6);
  });
});

describe('setPath', () => {
  it('sets a simple top-level key', () => {
    const obj = { a: 1 };
    const result = setPath(obj, 'b', 2);
    expect(result).toEqual({ a: 1, b: 2 });
  });

  it('sets a nested key and creates intermediates', () => {
    const obj = {};
    const result = setPath(obj, 'a.b.c', 'deep');
    expect(result).toEqual({ a: { b: { c: 'deep' } } });
  });

  it('sets an array index on an existing array', () => {
    const obj = { items: ['a', 'b', 'c'] };
    setPath(obj, 'items[1]', 'B');
    expect(obj.items).toEqual(['a', 'B', 'c']);
  });

  it('extends array with nulls when index is beyond length', () => {
    const obj = { items: ['a'] };
    setPath(obj, 'items[3]', 'D');
    expect(obj.items).toEqual(['a', null, null, 'D']);
  });

  it('creates array when next segment is index', () => {
    const obj: Record<string, unknown> = {};
    setPath(obj, 'list[0]', 'first');
    expect(obj['list']).toEqual(['first']);
  });

  it('on null root creates object', () => {
    const result = setPath(null, 'key', 'value');
    expect(result).toEqual({ key: 'value' });
  });

  it('on undefined root creates object', () => {
    const result = setPath(undefined, 'key', 'value');
    expect(result).toEqual({ key: 'value' });
  });

  it('throws on wildcard segment', () => {
    expect(() => setPath({ items: [1] }, 'items[*]', 99)).toThrow(/wildcard/);
  });

  it('throws on length segment', () => {
    expect(() => setPath({ items: [1] }, 'items.length', 5)).toThrow(/length/);
  });
});

describe('deletePath', () => {
  it('deletes a top-level key from object', () => {
    const obj = { a: 1, b: 2 };
    expect(deletePath(obj, 'a')).toBe(true);
    expect(obj).toEqual({ b: 2 });
  });

  it('deletes a nested key', () => {
    const obj = { a: { b: { c: 3 }, d: 4 } };
    expect(deletePath(obj, 'a.b.c')).toBe(true);
    expect(obj).toEqual({ a: { b: {}, d: 4 } });
  });

  it('deletes an array element by splicing', () => {
    const obj = { items: ['a', 'b', 'c'] };
    expect(deletePath(obj, 'items[1]')).toBe(true);
    expect(obj.items).toEqual(['a', 'c']);
  });

  it('returns false for missing key', () => {
    const obj = { a: 1 };
    expect(deletePath(obj, 'b')).toBe(false);
  });

  it('returns false on non-object root', () => {
    expect(deletePath('string', 'key')).toBe(false);
    expect(deletePath(42, 'key')).toBe(false);
    expect(deletePath(null, 'key')).toBe(false);
  });
});

File: utilities/shell/direnv-config/sdk/elixir/test/direnv_config/path_test.exs
defmodule DirenvConfig.PathTest do
  use ExUnit.Case, async: true

  alias DirenvConfig.Path

  describe "parse/1" do
    test "simple key" do
      assert Path.parse("name") == [{:key, "name"}]
    end

    test "dotted path" do
      assert Path.parse("a.b.c") == [{:key, "a"}, {:key, "b"}, {:key, "c"}]
    end

    test "index" do
      assert Path.parse("items[0]") == [{:key, "items"}, {:index, 0}]
    end

    test "negative index" do
      assert Path.parse("items[-1]") == [{:key, "items"}, {:index, -1}]
    end

    test "wildcard" do
      assert Path.parse("endpoints[*].host") == [
               {:key, "endpoints"},
               :wildcard,
               {:key, "host"}
             ]
    end

    test "length" do
      assert Path.parse("items.length") == [{:key, "items"}, :length]
    end

    test "chained brackets" do
      assert Path.parse("matrix[0][1]") == [{:key, "matrix"}, {:index, 0}, {:index, 1}]
    end

    test "mixed path" do
      assert Path.parse("folder[5].person.mobile") == [
               {:key, "folder"},
               {:index, 5},
               {:key, "person"},
               {:key, "mobile"}
             ]
    end

    test "empty string" do
      assert Path.parse("") == []
    end

    test "length as first token is a key" do
      assert Path.parse("length") == [{:key, "length"}]
    end
  end

  describe "get/2" do
    test "simple key" do
      assert Path.get(%{"name" => "alice"}, "name") == {:ok, "alice"}
    end

    test "nested dot" do
      data = %{"db" => %{"host" => "localhost", "port" => 5432}}
      assert Path.get(data, "db.host") == {:ok, "localhost"}
      assert Path.get(data, "db.port") == {:ok, 5432}
    end

    test "missing key" do
      assert Path.get(%{"a" => 1}, "b") == :error
    end

    test "array index" do
      data = %{"items" => ["alpha", "beta", "gamma"]}
      assert Path.get(data, "items[0]") == {:ok, "alpha"}
      assert Path.get(data, "items[2]") == {:ok, "gamma"}
    end

    test "negative index" do
      data = %{"items" => ["alpha", "beta", "gamma"]}
      assert Path.get(data, "items[-1]") == {:ok, "gamma"}
      assert Path.get(data, "items[-2]") == {:ok, "beta"}
    end

    test "out of bounds" do
      data = %{"items" => ["a"]}
      assert Path.get(data, "items[5]") == :error
      assert Path.get(data, "items[-5]") == :error
    end

    test "length of list" do
      data = %{"items" => ["a", "b", "c"]}
      assert Path.get(data, "items.length") == {:ok, 3}
    end

    test "length of map" do
      data = %{"m" => %{"a" => 1, "b" => 2}}
      assert Path.get(data, "m.length") == {:ok, 2}
    end

    test "wildcard" do
      data = %{
        "endpoints" => [
          %{"host" => "a.com", "port" => 80},
          %{"host" => "b.com", "port" => 443}
        ]
      }

      assert Path.get(data, "endpoints[*].host") == {:ok, ["a.com", "b.com"]}
    end

    test "chained index" do
      data = %{"matrix" => [[1, 2, 3], [4, 5, 6]]}
      assert Path.get(data, "matrix[0][1]") == {:ok, 2}
      assert Path.get(data, "matrix[1][-1]") == {:ok, 6}
    end

    test "deep nested" do
      data = %{"a" => %{"b" => %{"c" => "deep"}}}
      assert Path.get(data, "a.b.c") == {:ok, "deep"}
    end
  end

  describe "set/3" do
    test "set a simple top-level key" do
      assert Path.set(%{}, "name", "alice") == {:ok, %{"name" => "alice"}}
    end

    test "set a nested key creates intermediate maps" do
      assert Path.set(%{}, "a.b.c", 42) == {:ok, %{"a" => %{"b" => %{"c" => 42}}}}
    end

    test "set preserves existing keys" do
      data = %{"a" => %{"x" => 1}}
      assert Path.set(data, "a.y", 2) == {:ok, %{"a" => %{"x" => 1, "y" => 2}}}
    end

    test "set a list index" do
      data = %{"items" => ["a", "b", "c"]}
      assert Path.set(data, "items[1]", "B") == {:ok, %{"items" => ["a", "B", "c"]}}
    end

    test "set extends list with nils when index beyond length" do
      data = %{"items" => ["a"]}
      assert Path.set(data, "items[3]", "d") == {:ok, %{"items" => ["a", nil, nil, "d"]}}
    end

    test "set returns error on wildcard" do
      assert Path.set(%{"items" => [1, 2]}, "items[*]", 0) == {:error, :not_supported}
    end

    test "set returns error on length" do
      assert Path.set(%{"items" => [1, 2]}, "items.length", 5) == {:error, :not_supported}
    end
  end

  describe "delete/2" do
    test "delete a top-level key" do
      data = %{"a" => 1, "b" => 2}
      assert Path.delete(data, "a") == {:ok, %{"b" => 2}}
    end

    test "delete a nested key" do
      data = %{"a" => %{"b" => 1, "c" => 2}}
      assert Path.delete(data, "a.b") == {:ok, %{"a" => %{"c" => 2}}}
    end

    test "delete a list element" do
      data = %{"items" => ["a", "b", "c"]}
      assert Path.delete(data, "items[1]") == {:ok, %{"items" => ["a", "c"]}}
    end

    test "delete returns error for missing key" do
      data = %{"a" => 1}
      assert Path.delete(data, "b") == {:ok, %{"a" => 1}}
    end

    test "delete returns error for missing nested key" do
      data = %{"a" => %{"b" => 1}}
      assert Path.delete(data, "a.z") == {:ok, %{"a" => %{"b" => 1}}}
    end

    test "delete returns error for deeply missing path" do
      data = %{"a" => 1}
      assert Path.delete(data, "x.y.z") == :error
    end
  end
end
